File: f#/src/Sodium/Sodium.Tests/DenotationalSemanticsTests.fs
module Sodium.Tests.``Denotational Semantics``

open NUnit.Framework
open Sodium
open System.Collections.Generic

[<TestFixture>]
type Tests() =
    
    let runSimulationWithSimultaneousFirings listen firings =
        let maxKey =
            match firings with
            | [] -> -1
            | _ -> firings |> Seq.collect (Map.toSeq >> Seq.map fst) |> Seq.max
        let out = List<_>()
        let run t = for a in firings |> Seq.collect (fun f -> match f |> Map.tryFind t with | None -> [] | Some t -> t) do a ()
        match maxKey with
        | -1 ->
            use _l = listen out.Add
            ()
        | _ ->
            use _l = Transaction.Run (fun () ->
                let l = listen out.Add
                run 0
                l)
            for i = 1 to maxKey do Transaction.Run (fun () -> run i)
        out

    let runSimulationWithNoFirings listen =
        runSimulationWithSimultaneousFirings listen List.empty

    let runSimulation listen firings =
        runSimulationWithSimultaneousFirings listen (firings |> List.map (Map.map (fun _ a -> [a])))

    let mkStream firings =
        let s = Stream.sink ()
        let f = firings |> Map.map (fun _ v -> (fun () -> s.Send v))
        match f |> Map.tryFindKey (fun k _ -> k < 0) with
        | None -> ()
        | Some _ -> invalidOp "All firings must occur at T >= 0."
        (s :> 'a Stream, f)

    let mkStreamWithCoalesce firings coalesce =
        let s = Stream.sinkWithCoalesce coalesce
        let f = firings |> List.groupBy fst |> Map.ofList |> Map.map (fun _ v -> v |> (List.map (fun v -> (fun () -> s.Send (snd v)))))
        match f |> Map.tryFindKey (fun k _ -> k < 0) with
        | None -> ()
        | Some _ -> invalidOp "All firings must occur at T >= 0."
        (s :> 'a Stream, f)

    let rec getPermutationsInternal list length =
        if length = 1
        then list |> List.map (fun o -> [o])
        else getPermutationsInternal list (length - 1) |> List.collect (fun t -> list |> List.filter (fun e -> t |> List.contains e |> not) |> List.map (fun t2 -> List.append t [t2]))

    let getPermutations list = getPermutationsInternal list list.Length

    let runPermutations
        //(createListAndListener : ((string * Map<int, (unit -> unit)>) list -> (('T -> unit) -> IListener) -> (string * (int * (unit -> unit)) Map) list * (('T -> unit) -> IListener)) -> (string * Map<int, (unit -> unit)>) list * (('T -> unit) -> IListener))
        (createListAndListener : unit -> (string * Map<int, (unit -> unit)>) list * (('T -> unit) -> IListener))
        ``assert`` =
        let indexes = List.init (createListAndListener () |> fst).Length id
        for (list, listener) in (getPermutations indexes |> List.map (fun ii ->
            let (list, listener) = createListAndListener ()
            (ii |> List.map (fun i -> list.[i]), listener))) do
            try
                let out = runSimulation listener (list |> List.map snd)
                ``assert`` out
            with
                | _ ->
                    printfn "Test failed for ordering { %s }." (list |> List.map fst |> String.concat ", ")
                    reraise ()

    [<Test>]
    member __.``Never: Test Case``() =
        let out = runSimulationWithNoFirings (Helper.flip Stream.listen (Stream.never<int> ()))
        CollectionAssert.AreEqual([], out)

    [<Test>]
    member __.``MapS: Test Case``() =
        let (s, sf) = mkStream ([(0,5);(1,10);(2,12)] |> Map.ofList)
        let out = runSimulation (Helper.flip Stream.listen (s |> Stream.map ((+) 1))) [sf]
        CollectionAssert.AreEqual([6;11;13], out)

    [<Test>]
    member __.``Snapshot: Test Case``() =
        let (s1, s1f) = mkStream ([(0,'a');(3,'b');(5,'c')] |> Map.ofList)
        let (s2, s2f) = mkStream ([(1,4);(5,7)] |> Map.ofList)
        let c = s2 |> Stream.hold 3
        let out = runSimulation (Helper.flip Stream.listen (s1 |> Stream.snapshotAndTakeCell c)) [s1f;s2f]
        CollectionAssert.AreEqual([3;4;4], out)

    [<Test>]
    member __.``Merge: Test Case``() =
        let (s1, s1f) = mkStream ([(0,0);(2,2)] |> Map.ofList)
        let (s2, s2f) = mkStream ([(1,10);(2,20);(3,30)] |> Map.ofList)
        let out = runSimulation (Helper.flip Stream.listen (s1 |> Stream.merge (+) s2)) [s1f;s2f]
        CollectionAssert.AreEqual([0;10;22;30], out)

    [<Test>]
    member __.``Filter: Test Case``() =
        let (s, sf) = mkStream ([(0,5);(1,6);(2,7)] |> Map.ofList)
        let out = runSimulation (Helper.flip Stream.listen (s |> Stream.filter ((Helper.flip (%) 2) >> ((<>) 0)))) [sf]
        CollectionAssert.AreEqual([5;7], out)

    [<Test>]
    member __.``SwitchS: Test Case``() =
        runPermutations
            (fun () ->
                let (s1, s1f) = mkStream ([(0,'a');(1,'b');(2,'c');(3,'d')] |> Map.ofList)
                let (s2, s2f) = mkStream ([(0,'W');(1,'X');(2,'Y');(3,'Z')] |> Map.ofList)
                let (switcher, switcherF) = mkStream ([1,s2] |> Map.ofList)
                let c = switcher |> Stream.hold s1
                let firings = [("s1",s1f);("s2",s2f);("switcher",switcherF)]
                (firings, Helper.flip Stream.listen (c |> Cell.switchS)))
            (fun out -> CollectionAssert.AreEqual(['a';'b';'Y';'Z'], out))

    [<Test>]
    member __.``Updates: Test Case``() =
        let (s, sf) = mkStream ([(1,'b');(3,'c')] |> Map.ofList)
        let c = s |> Stream.hold 'a'
        let out = runSimulation (Helper.flip Stream.listen (c |> Operational.updates)) [sf]
        CollectionAssert.AreEqual(['b';'c'], out)

    [<Test>]
    member __.``Value: Test Case 1``() =
        let (s, sf) = mkStream ([(1,'b');(3,'c')] |> Map.ofList)
        let c = s |> Stream.hold 'a'
        let out = runSimulation (fun h -> Transaction.Run (fun () -> c |> Operational.value |> Stream.listen h)) [sf]
        CollectionAssert.AreEqual(['a';'b';'c'], out)

    [<Test>]
    member __.``Value: Test Case 2``() =
        let (s, sf) = mkStream ([(0,'b');(1,'c');(3,'d')] |> Map.ofList)
        let c = s |> Stream.hold 'a'
        let out = runSimulation (fun h -> Transaction.Run (fun () -> c |> Operational.value |> Stream.listen h)) [sf]
        CollectionAssert.AreEqual(['b';'c';'d'], out)

    [<Test>]
    member __.``ListenC: Test Case 1``() =
        let (s, sf) = mkStream ([(1,'b');(3,'c')] |> Map.ofList)
        let c = s |> Stream.hold 'a'
        let out = runSimulation (Helper.flip Cell.listen c) [sf]
        CollectionAssert.AreEqual(['a';'b';'c'], out)

    [<Test>]
    member __.``ListenC: Test Case 2``() =
        let (s, sf) = mkStream ([(0,'b');(1,'c');(3,'d')] |> Map.ofList)
        let c = s |> Stream.hold 'a'
        let out = runSimulation (Helper.flip Cell.listen c) [sf]
        CollectionAssert.AreEqual(['b';'c';'d'], out)

    [<Test>]
    member __.``Split: Test Case``() =
        let (s, sf) = mkStreamWithCoalesce [(0,['a';'b']);(1,['c']);(1,['d';'e'])] List.append
        let out = runSimulationWithSimultaneousFirings (Helper.flip Stream.listen (s |> Operational.split)) [sf]
        CollectionAssert.AreEqual(['a';'b';'c';'d';'e'], out)

    [<Test>]
    member __.``Constant: Test Case``() =
        let c = Cell.constant 'a'
        let out = runSimulationWithNoFirings (Helper.flip Cell.listen c)
        CollectionAssert.AreEqual(['a'], out)

    [<Test>]
    member __.``ConstantLazy: Test Case``() =
        let c = Cell.constantLazy (lazy 'a')
        let out = runSimulationWithNoFirings (Helper.flip Cell.listen c)
        CollectionAssert.AreEqual(['a'], out)

    [<Test>]
    member __.``Hold: Test Case``() =
        let (s, sf) = mkStream ([(1,'b');(3,'c')] |> Map.ofList)
        let c = s |> Stream.hold 'a'
        let out = runSimulation (Helper.flip Cell.listen c) [sf]
        CollectionAssert.AreEqual(['a';'b';'c'], out)

    [<Test>]
    member __.``MapC: Test Case``() =
        let (s, sf) = mkStream ([(2,3);(3,5)] |> Map.ofList)
        let c = s |> Stream.hold 0
        let out = runSimulation (Helper.flip Cell.listen (c |> Cell.map ((+) 1))) [sf]
        CollectionAssert.AreEqual([1;4;6], out)

    [<Test>]
    member __.``Apply: Test Case``() =
        let (s1, s1f) = mkStream ([(1,200);(2,300);(4,400)] |> Map.ofList)
        let ca = s1 |> Stream.hold 100
        let (s2, s2f) = mkStream ([(1,(+) 5);(3,(+) 6)] |> Map.ofList)
        let cf = s2 |> Stream.hold ((+) 0)
        let out = runSimulation (Helper.flip Cell.listen (ca |> Cell.apply cf)) [s1f;s2f]
        CollectionAssert.AreEqual([100;205;305;306;406], out)

    [<Test>]
    member __.``SwitchC: Test Case 1``() =
        runPermutations
            (fun () ->
                let (s1, s1f) = mkStream ([(0,'b');(1,'c');(2,'d');(3,'e')] |> Map.ofList)
                let c1 = s1 |> Stream.hold 'a'
                let (s2, s2f) = mkStream ([(0,'W');(1,'X');(2,'Y');(3,'Z')] |> Map.ofList)
                let c2 = s2 |> Stream.hold 'V'
                let (switcher, switcherF) = mkStream ([1,c2] |> Map.ofList)
                let c = switcher |> Stream.hold c1
                let firings = [("s1",s1f);("s2",s2f);("switcher",switcherF)]
                (firings, Helper.flip Cell.listen (c |> Cell.switchC)))
            (fun out -> CollectionAssert.AreEqual(['b';'X';'Y';'Z'], out))

    [<Test>]
    member __.``SwitchC: Test Case 2``() =
        runPermutations
            (fun () ->
                let (s1, s1f) = mkStream ([(0,'b');(1,'c');(2,'d');(3,'e')] |> Map.ofList)
                let c1 = s1 |> Stream.hold 'a'
                let (s2, s2f) = mkStream ([(1,'X');(2,'Y');(3,'Z')] |> Map.ofList)
                let c2 = s2 |> Stream.hold 'W'
                let (switcher, switcherF) = mkStream ([1,c2] |> Map.ofList)
                let c = switcher |> Stream.hold c1
                let firings = [("s1",s1f);("s2",s2f);("switcher",switcherF)]
                (firings, Helper.flip Cell.listen (c |> Cell.switchC)))
            (fun out -> CollectionAssert.AreEqual(['b';'X';'Y';'Z'], out))

    [<Test>]
    member __.``SwitchC: Test Case 3``() =
        runPermutations
            (fun () ->
                let (s1, s1f) = mkStream ([(0,'b');(1,'c');(2,'d');(3,'e')] |> Map.ofList)
                let c1 = s1 |> Stream.hold 'a'
                let (s2, s2f) = mkStream ([(2,'Y');(3,'Z')] |> Map.ofList)
                let c2 = s2 |> Stream.hold 'X'
                let (switcher, switcherF) = mkStream ([1,c2] |> Map.ofList)
                let c = switcher |> Stream.hold c1
                let firings = [("s1",s1f);("s2",s2f);("switcher",switcherF)]
                (firings, Helper.flip Cell.listen (c |> Cell.switchC)))
            (fun out -> CollectionAssert.AreEqual(['b';'X';'Y';'Z'], out))

    [<Test>]
    member __.``SwitchC: Test Case 4``() =
        runPermutations
            (fun () ->
                let (s1, s1f) = mkStream ([(0,'b');(1,'c');(2,'d');(3,'e')] |> Map.ofList)
                let c1 = s1 |> Stream.hold 'a'
                let (s2, s2f) = mkStream ([(0,'W');(1,'X');(2,'Y');(3,'Z')] |> Map.ofList)
                let c2 = s2 |> Stream.hold 'V'
                let (s3, s3f) = mkStream ([(0,'2');(1,'3');(2,'4');(3,'5')] |> Map.ofList)
                let c3 = s3 |> Stream.hold '1'
                let (switcher, switcherF) = mkStream ([(1,c2);(3,c3)] |> Map.ofList)
                let c = switcher |> Stream.hold c1
                let firings = [("s1",s1f);("s2",s2f);("s3",s3f);("switcher",switcherF)]
                (firings, Helper.flip Cell.listen (c |> Cell.switchC)))
            (fun out -> CollectionAssert.AreEqual(['b';'X';'Y';'5'], out))

    [<Test>]
    member __.``Sample: Test Case``() =
        let s = Stream.sink ()
        let c = s |> Stream.hold 'a'
        let sample1 = c |> Cell.sample
        s.Send 'b'
        let sample2 = c |> Cell.sample
        Assert.AreEqual('a', sample1)
        Assert.AreEqual('b', sample2)

    [<Test>]
    member __.``SampleLazy: Test Case``() =
        let s = Stream.sink ()
        let c = s |> Stream.hold 'a'
        let sample1 = c |> Cell.sampleLazy
        s.Send 'b'
        let sample2 = c |> Cell.sampleLazy
        Assert.AreEqual('a', sample1.Value)
        Assert.AreEqual('b', sample2.Value)

File: dotnet/src/Sodium/CSharp/Sodium.Frp.Tests/DenotationalSemanticsTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using NUnit.Framework;

namespace Sodium.Frp.Tests
{
    [TestFixture]
    public class DenotationalSemanticsTests
    {
        [Test]
        public void Test_Never_TestCase()
        {
            List<int> @out = RunSimulation<int>(Stream.Never<int>().Listen);
            CollectionAssert.AreEqual(new int[0], @out);
        }

        [Test]
        public void Test_MapS_TestCase()
        {
            (Stream<int> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, int> { { 0, 5 }, { 1, 10 }, { 2, 12 } });
            List<int> @out = RunSimulation<int>(s.Map(x => x + 1).Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 6, 11, 13 }, @out);
        }

        [Test]
        public void Test_Snapshot_TestCase()
        {
            (Stream<char> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, char> { { 0, 'a' }, { 3, 'b' }, { 5, 'c' } });
            (Stream<int> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, int> { { 1, 4 }, { 5, 7 } });
            Cell<int> c = s2.Hold(3);
            List<int> @out = RunSimulation<int>(s1.Snapshot(c).Listen, new[] { s1F, s2F });
            CollectionAssert.AreEqual(new[] { 3, 4, 4 }, @out);
        }

        [Test]
        public void Test_Merge_TestCase()
        {
            (Stream<int> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, int> { { 0, 0 }, { 2, 2 } });
            (Stream<int> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, int> { { 1, 10 }, { 2, 20 }, { 3, 30 } });
            List<int> @out = RunSimulation<int>(s1.Merge(s2, (x, y) => x + y).Listen, new[] { s1F, s2F });
            CollectionAssert.AreEqual(new[] { 0, 10, 22, 30 }, @out);
        }

        [Test]
        public void Test_Filter_TestCase()
        {
            (Stream<int> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, int> { { 0, 5 }, { 1, 6 }, { 2, 7 } });
            List<int> @out = RunSimulation<int>(s.Filter(x => x % 2 != 0).Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 5, 7 }, @out);
        }

        [Test]
        public void Test_SwitchS_TestCase()
        {
            RunPermutations<char>(
                createFiringsListAndListener =>
                {
                    (Stream<char> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, char> { { 0, 'a' }, { 1, 'b' }, { 2, 'c' }, { 3, 'd' } });
                    (Stream<char> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, char> { { 0, 'W' }, { 1, 'X' }, { 2, 'Y' }, { 3, 'Z' } });
                    (Stream<Stream<char>> switcher, Dictionary<int, Action> switcherF) = MkStream(new Dictionary<int, Stream<char>> { { 1, s2 } });
                    Cell<Stream<char>> c = switcher.Hold(s1);

                    IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> firings = new[]
                    {
                        (Name: "s1", Firings: s1F),
                        (Name: "s2", Firings: s2F),
                        (Name: "switcher", Firings: switcherF)
                    };

                    return createFiringsListAndListener(firings, c.SwitchS().Listen);
                },
                @out => CollectionAssert.AreEqual(new[] { 'a', 'b', 'Y', 'Z' }, @out));
        }

        [Test]
        public void Test_Updates_TestCase()
        {
            (Stream<char> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, char> { { 1, 'b' }, { 3, 'c' } });
            Cell<char> c = s.Hold('a');
            List<char> @out = RunSimulation<char>(c.Updates().Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 'b', 'c' }, @out);
        }

        [Test]
        public void Test_Value_TestCase1()
        {
            (Stream<char> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, char> { { 1, 'b' }, { 3, 'c' } });
            Cell<char> c = s.Hold('a');
            List<char> @out = RunSimulation<char>(h => Transaction.Run(() => c.Values().Listen(h)), new[] { sf });
            CollectionAssert.AreEqual(new[] { 'a', 'b', 'c' }, @out);
        }

        [Test]
        public void Test_Value_TestCase2()
        {
            (Stream<char> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, char> { { 0, 'b' }, { 1, 'c' }, { 3, 'd' } });
            Cell<char> c = s.Hold('a');
            List<char> @out = RunSimulation<char>(h => Transaction.Run(() => c.Values().Listen(h)), new[] { sf });
            CollectionAssert.AreEqual(new[] { 'b', 'c', 'd' }, @out);
        }

        [Test]
        public void Test_ListenC_TestCase1()
        {
            (Stream<char> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, char> { { 1, 'b' }, { 3, 'c' } });
            Cell<char> c = s.Hold('a');
            List<char> @out = RunSimulation<char>(c.Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 'a', 'b', 'c' }, @out);
        }

        [Test]
        public void Test_ListenC_TestCase2()
        {
            (Stream<char> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, char> { { 0, 'b' }, { 1, 'c' }, { 3, 'd' } });
            Cell<char> c = s.Hold('a');
            List<char> @out = RunSimulation<char>(c.Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 'b', 'c', 'd' }, @out);
        }

        [Test]
        public void Test_Split_TestCase()
        {
            (Stream<IReadOnlyList<char>> s, ILookup<int, Action> sf) = MkStream(
                new(int Time, IReadOnlyList<char> Value)[]
                {
                    (Time: 0, Value: new[] { 'a', 'b' }),
                    (Time: 1, Value: new[] { 'c' }),
                    (Time: 1, Value: new[] { 'd', 'e' })
                },
                (x, y) => x.Concat(y).ToArray());
            List<char> @out = RunSimulation<char>(Operational.Split<char, IReadOnlyList<char>>(s).Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 'a', 'b', 'c', 'd', 'e' }, @out);
        }

        [Test]
        public void Test_Constant_TestCase()
        {
            Cell<char> c = Cell.Constant('a');
            List<char> @out = RunSimulation<char>(c.Listen);
            CollectionAssert.AreEqual(new[] { 'a' }, @out);
        }

        [Test]
        public void Test_ConstantLazy_TestCase()
        {
            Cell<char> c = Cell.ConstantLazy(new Lazy<char>(() => 'a'));
            List<char> @out = RunSimulation<char>(c.Listen);
            CollectionAssert.AreEqual(new[] { 'a' }, @out);
        }

        [Test]
        public void Test_Hold_TestCase()
        {
            (Stream<char> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, char> { { 1, 'b' }, { 3, 'c' } });
            Cell<char> c = s.Hold('a');
            List<char> @out = RunSimulation<char>(c.Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 'a', 'b', 'c' }, @out);
        }

        [Test]
        public void Test_MapC_TestCase()
        {
            (Stream<int> s, Dictionary<int, Action> sf) = MkStream(new Dictionary<int, int> { { 2, 3 }, { 3, 5 } });
            Cell<int> c = s.Hold(0);
            List<int> @out = RunSimulation<int>(c.Map(x => x + 1).Listen, new[] { sf });
            CollectionAssert.AreEqual(new[] { 1, 4, 6 }, @out);
        }

        [Test]
        public void Test_Apply_TestCase()
        {
            (Stream<int> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, int> { { 1, 200 }, { 2, 300 }, { 4, 400 } });
            Cell<int> ca = s1.Hold(100);
            (Stream<Func<int, int>> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, Func<int, int>> { { 1, x => x + 5 }, { 3, x => x + 6 } });
            Cell<Func<int, int>> cf = s2.Hold(x => x + 0);
            List<int> @out = RunSimulation<int>(ca.Apply(cf).Listen, new[] { s1F, s2F });
            CollectionAssert.AreEqual(new[] { 100, 205, 305, 306, 406 }, @out);
        }

        [Test]
        public void Test_SwitchC_TestCase1()
        {
            RunPermutations<char>(createFiringsListAndListener =>
                {
                    (Stream<char> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, char> { { 0, 'b' }, { 1, 'c' }, { 2, 'd' }, { 3, 'e' } });
                    Cell<char> c1 = s1.Hold('a');
                    (Stream<char> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, char> { { 0, 'W' }, { 1, 'X' }, { 2, 'Y' }, { 3, 'Z' } });
                    Cell<char> c2 = s2.Hold('V');
                    (Stream<Cell<char>> switcher, Dictionary<int, Action> switcherF) = MkStream(new Dictionary<int, Cell<char>> { { 1, c2 } });
                    Cell<Cell<char>> c = switcher.Hold(c1);

                    IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> firings = new[]
                    {
                        (Name: "s1", Firings: s1F),
                        (Name: "s2", Firings: s2F),
                        (Name: "switcher", Firings: switcherF)
                    };

                    return createFiringsListAndListener(firings, c.SwitchC().Listen);
                },
                @out => CollectionAssert.AreEqual(new[] { 'b', 'X', 'Y', 'Z' }, @out));
        }

        [Test]
        public void Test_SwitchC_TestCase2()
        {
            RunPermutations<char>(createFiringsListAndListener =>
                {
                    (Stream<char> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, char> { { 0, 'b' }, { 1, 'c' }, { 2, 'd' }, { 3, 'e' } });
                    Cell<char> c1 = s1.Hold('a');
                    (Stream<char> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, char> { { 1, 'X' }, { 2, 'Y' }, { 3, 'Z' } });
                    Cell<char> c2 = s2.Hold('W');
                    (Stream<Cell<char>> switcher, Dictionary<int, Action> switcherF) = MkStream(new Dictionary<int, Cell<char>> { { 1, c2 } });
                    Cell<Cell<char>> c = switcher.Hold(c1);

                    IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> firings = new[]
                    {
                        (Name: "s1", Firings: s1F),
                        (Name: "s2", Firings: s2F),
                        (Name: "switcher", Firings: switcherF)
                    };

                    return createFiringsListAndListener(firings, c.SwitchC().Listen);
                },
                @out => CollectionAssert.AreEqual(new[] { 'b', 'X', 'Y', 'Z' }, @out));
        }

        [Test]
        public void Test_SwitchC_TestCase3()
        {
            RunPermutations<char>(createFiringsListAndListener =>
                {
                    (Stream<char> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, char> { { 0, 'b' }, { 1, 'c' }, { 2, 'd' }, { 3, 'e' } });
                    Cell<char> c1 = s1.Hold('a');
                    (Stream<char> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, char> { { 2, 'Y' }, { 3, 'Z' } });
                    Cell<char> c2 = s2.Hold('X');
                    (Stream<Cell<char>> switcher, Dictionary<int, Action> switcherF) = MkStream(new Dictionary<int, Cell<char>> { { 1, c2 } });
                    Cell<Cell<char>> c = switcher.Hold(c1);

                    IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> firings = new[]
                    {
                        (Name: "s1", Firings: s1F),
                        (Name: "s2", Firings: s2F),
                        (Name: "switcher", Firings: switcherF)
                    };

                    return createFiringsListAndListener(firings, c.SwitchC().Listen);
                },
                @out => CollectionAssert.AreEqual(new[] { 'b', 'X', 'Y', 'Z' }, @out));
        }

        [Test]
        public void Test_SwitchC_TestCase4()
        {
            RunPermutations<char>(createFiringsListAndListener =>
                {
                    (Stream<char> s1, Dictionary<int, Action> s1F) = MkStream(new Dictionary<int, char> { { 0, 'b' }, { 1, 'c' }, { 2, 'd' }, { 3, 'e' } });
                    Cell<char> c1 = s1.Hold('a');
                    (Stream<char> s2, Dictionary<int, Action> s2F) = MkStream(new Dictionary<int, char> { { 0, 'W' }, { 1, 'X' }, { 2, 'Y' }, { 3, 'Z' } });
                    Cell<char> c2 = s2.Hold('V');
                    (Stream<char> s3, Dictionary<int, Action> s3F) = MkStream(new Dictionary<int, char> { { 0, '2' }, { 1, '3' }, { 2, '4' }, { 3, '5' } });
                    Cell<char> c3 = s3.Hold('1');
                    (Stream<Cell<char>> switcher, Dictionary<int, Action> switcherF) = MkStream(new Dictionary<int, Cell<char>> { { 1, c2 }, { 3, c3 } });
                    Cell<Cell<char>> c = switcher.Hold(c1);

                    IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> firings = new[]
                    {
                        (Name: "s1", Firings: s1F),
                        (Name: "s2", Firings: s2F),
                        (Name: "s3", Firings: s3F),
                        (Name: "switcher", Firings: switcherF)
                    };

                    return createFiringsListAndListener(firings, c.SwitchC().Listen);
                },
                @out => CollectionAssert.AreEqual(new[] { 'b', 'X', 'Y', '5' }, @out));
        }

        [Test]
        public void Test_Sample_TestCase()
        {
            StreamSink<char> s = Stream.CreateSink<char>();
            Cell<char> c = s.Hold('a');
            char sample1 = c.Sample();
            s.Send('b');
            char sample2 = c.Sample();
            Assert.AreEqual('a', sample1);
            Assert.AreEqual('b', sample2);
        }

        [Test]
        public void Test_SampleLazy_TestCase()
        {
            StreamSink<char> s = Stream.CreateSink<char>();
            Cell<char> c = s.Hold('a');
            Lazy<char> sample1 = c.SampleLazy();
            s.Send('b');
            Lazy<char> sample2 = c.SampleLazy();
            Assert.AreEqual('a', sample1.Value);
            Assert.AreEqual('b', sample2.Value);
        }

        private static (Stream<T> Stream, Dictionary<int, Action> Firings) MkStream<T>(Dictionary<int, T> firings)
        {
            StreamSink<T> s = Stream.CreateSink<T>();
            Dictionary<int, Action> f = firings.ToDictionary(firing => firing.Key, firing => (Action)(() => s.Send(firing.Value)));
            if (f.Keys.Any(k => k < 0))
            {
                throw new InvalidOperationException("All firings must occur at T >= 0.");
            }
            Stream<T> returnStream = s;
            return (Stream: returnStream, Firings: f);
        }

        private static (Stream<T> Stream, ILookup<int, Action> Firings) MkStream<T>(IReadOnlyList<(int Time, T Value)> firings, Func<T, T, T> coalesce)
        {
            StreamSink<T> s = Stream.CreateSink(coalesce);
            ILookup<int, Action> f = firings.ToLookup(firing => firing.Time, firing => (Action)(() => s.Send(firing.Value)));
            if (f.Any(g => g.Key < 0))
            {
                throw new InvalidOperationException("All firings must occur at T >= 0.");
            }
            Stream<T> returnStream = s;
            return (Stream: returnStream, Firings: f);
        }

        private static List<T> RunSimulation<T>(Func<Action<T>, IListener> listen, IReadOnlyList<Dictionary<int, Action>> firings)
        {
            return RunSimulation(listen, firings.Select(f => f.ToLookup(p => p.Key, p => p.Value)).ToArray());
        }

        private static List<T> RunSimulation<T>(Func<Action<T>, IListener> listen, IReadOnlyList<ILookup<int, Action>> firings = null)
        {
            int maxKey = firings?.SelectMany(d => d.Select(g => g.Key)).DefaultIfEmpty(-1).Max() ?? -1;
            List<T> @out = new List<T>();
            IListener l = null;
            try
            {
                void Run(int t)
                {
                    if (firings != null)
                    {
                        foreach (Action a in firings.SelectMany(f => f[t]))
                        {
                            a();
                        }
                    }
                }

                if (maxKey > -1)
                {
                    l = Transaction.Run(() =>
                    {
                        IListener lLocal = listen(@out.Add);
                        Run(0);
                        return lLocal;
                    });

                    for (int i = 1; i <= maxKey; i++)
                    {
                        int t = i;
                        Transaction.RunVoid(() => { Run(t); });
                    }
                }
                else
                {
                    l = listen(@out.Add);
                }
            }
            finally
            {
                l?.Unlisten();
            }
            return @out;
        }

        private static void RunPermutations<T>(Func<Func<IReadOnlyList<(string Name, Dictionary<int, Action> Firings)>, Func<Action<T>, IListener>, (IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> FiringsList, Func<Action<T>, IListener> Listen)>, (IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> FiringsList, Func<Action<T>, IListener> Listen)> createListAndListener, Action<IReadOnlyList<T>> assert)
        {
            IReadOnlyList<int> indexes = Enumerable.Range(0, createListAndListener((fl, l) => (FiringsList: fl, Listen: l)).FiringsList.Count).ToArray();
            foreach ((IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> firingsList, Func<Action<T>, IListener> listener) in
                GetPermutations(indexes).Select(ii =>
                {
                    (IReadOnlyList<(string Name, Dictionary<int, Action> Firings)> firingsList, Func<Action<T>, IListener> listen) = createListAndListener((fl, l) => (FiringsList: fl, Listen: l));
                    return (FiringsList: ii.Select(i => firingsList[i]).ToArray(), Listen: listen);
                }))
            {
                try
                {
                    List<T> @out = RunSimulation(listener, firingsList.Select(o => o.Firings).ToArray());
                    assert(@out);
                }
                catch
                {
                    Console.WriteLine("Test failed for ordering { " + string.Join(", ", firingsList.Select(o => o.Name)) + " }.");
                    throw;
                }
            }
        }

        private static IReadOnlyList<IReadOnlyList<T>> GetPermutations<T>(IReadOnlyList<T> list)
        {
            return GetPermutations(list, list.Count);
        }

        private static IReadOnlyList<IReadOnlyList<T>> GetPermutations<T>(IReadOnlyList<T> list, int length)
        {
            if (length == 1)
            {
                return list.Select(t => new[] { t }).ToArray();
            }

            return GetPermutations(list, length - 1).SelectMany(t => list.Where(e => !t.Contains(e)), (t1, t2) => t1.Concat(new[] { t2 }).ToArray()).ToArray();
        }
    }
}